sdk/storage/azure_storage_blob/src/partitioned_transfer/upload.rs

```diff
@@ -0,0 +1,349 @@
+use azure_core::http::Body;
+use bytes::Bytes;
+
+#[cfg(not(target_arch = "wasm32"))]
+use azure_core::stream::SeekableStream;
+#[cfg(not(target_arch = "wasm32"))]
+use futures::StreamExt;
+
+#[cfg(not(target_arch = "wasm32"))]
+use crate::streams::partitioned_stream::PartitionedStream;
+
+use super::*;
+
+#[async_trait::async_trait]
+pub(crate) trait PartitionedUploadBehavior {
+    async fn transfer_oneshot(&self, content: Body) -> AzureResult<()>;
+    async fn transfer_partition(&self, offset: usize, content: Body) -> AzureResult<()>;
+    async fn initialize(&self, content_len: usize) -> AzureResult<()>;
+    async fn finalize(&self) -> AzureResult<()>;
+}
+
+pub(crate) async fn upload(
```

Comment:
I really love how all the abstractions let us have such a clear and simple definition of the algorithm like this. This looks great.